src/android/org/uproxy/tun2socks/DnsResolverService.java

```diff
@@ -0,0 +1,285 @@
+
+import android.app.Service;
+import android.content.Intent;
+import android.os.Binder;
+import android.os.IBinder;
+import android.support.v4.content.LocalBroadcastManager;
+import android.util.Log;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.nio.ByteBuffer;
+
+import socks.Socks5Proxy;
+import socks.SocksSocket;
+
+// This class listens on a UDP socket for incoming DNS traffic, which is proxy'd
+// through SOCKS over TCP to Google's Public DNS.
+public class DnsResolverService extends Service {
+
+  private static final String LOG_TAG = "DnsResolverService";
+  public static final String DNS_ADDRESS_BROADCAST = "dnsResolverAddressBroadcast";
+  public static final String DNS_ADDRESS_EXTRA = "dnsResolverAddress";
+  public static final String SOCKS_SERVER_ADDRESS_EXTRA = "socksServerAddress";
+
+  private final IBinder binder = new LocalBinder();
+  private String m_socksServerAddress;
+  private DnsUdpToSocksResolver dnsResolver = null;
+
+  public class LocalBinder extends Binder {
+    public DnsResolverService getService() {
+      return DnsResolverService.this;
+    }
+  }
+
+  @Override
+  public IBinder onBind(Intent intent) {
+    return binder;
+  }
+
+  @Override
+  public int onStartCommand(Intent intent, int flags, int startId) {
+    Log.i(LOG_TAG, "start command");
+    m_socksServerAddress = intent.getStringExtra(SOCKS_SERVER_ADDRESS_EXTRA);
+    if (m_socksServerAddress == null) {
+      Log.e(LOG_TAG, "Failed to receive socks server address.");
+    }
+    return START_NOT_STICKY;
+  }
+
+  @Override
+  public void onCreate() {
+    Log.i(LOG_TAG, "create");
+    dnsResolver = new DnsUdpToSocksResolver(DnsResolverService.this);
+    dnsResolver.start();
+  }
+
+  @Override
+  public void onDestroy() {
+    Log.i(LOG_TAG, "destroy");
+    if (dnsResolver != null) {
+      dnsResolver.interrupt();
+    }
+  }
+
+  // Parses the socks server address received on start and returns it as
+  // a InetSocketAddress.
+  public InetSocketAddress getSocksServerAddress() {
+    if (m_socksServerAddress == null) {
+      return null;
+    }
+    int separatorIndex = m_socksServerAddress.indexOf(':');
+    String ip = m_socksServerAddress.substring(0, separatorIndex);
+    int port = Integer.parseInt(m_socksServerAddress.substring(separatorIndex + 1));
+    return new InetSocketAddress(ip, port);
+  }
+
+  private class DnsUdpToSocksResolver extends Thread {
+    private static final String LOG_TAG = "DnsUdpToSocksResolver";
+    private static final String DNS_RESOLVER_IP = "8.8.8.8";
+    private static final int MAX_UDP_DATAGRAM_LEN = 32767; // DNS max over UDP
```

Comment:
Nit: Based on our conversation today, it sounds like the theoretical max is 64K, and the realistic max is 8K.  I'm fine with 32K too, but I think we should improve the comment regardless.